magazine.c

```diff
@@ -8,45 +8,165 @@
+
+  if (mag_state.mag_events < 2) {
+    return false;
+  }
+
+  return current_event->direction != last_event->direction;
+}
+
+void magazine_calculate_loss(void)
+{
+  const int last_evt_no = mag_state.evt_no ? 0 : 1;
+  const struct probe_window *current_event = &mag_state.probe_events[mag_state.evt_no];
+  const struct probe_window *last_event = &mag_state.probe_events[last_evt_no];
+
+  // Select which slop type we can detect based on whether or not lash
+  // is possible
+  int32_t * const property = magazine_lash_possible() ?
+    &mag_state.lash : &mag_state.spacing;
+
+  // Which edge we use as a reference point for our slop calculation
+  // depends on which direction we are moving in.
+  if (magazine_direction()) {
+    *property = abs(current_event->outboard - last_event->outboard);
+  } else {
+    *property = abs(current_event->inboard - last_event->inboard);
+  }
+}
+
+void magazine_record_edge(enum magazine_edge_type edge)
+{
+  // Update our edge events based on which direction we are travelling
+  // in, and whether or not it is a rising or falling edge
+  struct probe_window *current_event = &mag_state.probe_events[mag_state.evt_no];
+  if (magazine_direction() ^ (E_MAGAZINE_EDGE_TYPE_RISING == edge)) {
+    current_event->inboard = sys.position[C_AXIS];
+  } else {
+    current_event->outboard = sys.position[C_AXIS];
+  }
+}
+
+void magazine_loss_detector(const uint8_t magazine_alignment_on)
+{
+  // When the probe is detected, copy the current carousel position
+  // into the probe position. If we were not previously on a flag,
+  // begin recording a magazine flag event, otherwise, end the event
+  // and calculate the loss
+  if (magazine_alignment_on && !mag_state.on_probe) {
+    mag_state.on_probe = true;
+    mag_state.probe_events[mag_state.evt_no].direction = magazine_direction();
+    magazine_record_edge(E_MAGAZINE_EDGE_TYPE_RISING);
+    mag_state.mag_events++;
+  } else if (mag_state.on_probe) {
+    mag_state.on_probe = false;
+    magazine_record_edge(E_MAGAZINE_EDGE_TYPE_FALLING);
+    mag_state.evt_no = mag_state.evt_no ? 0 : 1;
+
+    magazine_calculate_loss();
+    request_report(REQUEST_SLOP_REPORT, 0);
+  }
+
 }
 
 //KeyMe function
 // Monitors the gap in units between mags and throws an alarm if the gap is larger than a
-// specified threshold.
+// specified threshold. Additionally, magazine slop can be calculated
+// if turned on via compiler flag
 void magazine_gap_monitor()
 {
   const uint8_t magazine_alignment_on = magazine_get_state();
-  // When the probe is detected, copy the current position of all axes
-  // into sys.probe_position
-  if(magazine_alignment_on){
-    memcpy(sys.probe_position, sys.position, sizeof(int32_t) * N_AXIS);
+  // When the probe is detected, copy the current carousel position
+  // into the probe position
+  if (magazine_alignment_on) {
+    sys.probe_position[C_AXIS] = sys.position[C_AXIS];
```

Comment:
No. That's insignificant to the point of not being worth it. As we're already discussing potentially adding probes on the x and y axes, I'd rather leave this in.